src/core/Coroutine/Barrier.php

```diff
@@ -27,13 +27,15 @@ public function __destruct()
     {
         if ($this->timer != -1) {
             Timer::clear($this->timer);
-            if (static::$cancel_list[$this->cid]) {
+            if (isset(static::$cancel_list[$this->cid])) {
                 unset(static::$cancel_list[$this->cid]);
                 return;
             }
         }
-        if ($this->cid != -1) {
+        if ($this->cid != -1 && $this->cid != Coroutine::getCid()) {
             Coroutine::resume($this->cid);
+        } else {
+            static::$cancel_list[$this->cid] = true;
```

Comment:
Line 38 should better be:

```php
            if (isset(static::$cancel_list[$this->cid])) {
                unset(static::$cancel_list[$this->cid]);
            }
```

Reply:
I'm actually waiting for a review from @matyhtf before the merge. 😂

If we use this, then the current coroutine result will be lost, Can't output `int(4)
int(4)
`.

```php
run(function () {
    $barrier = Barrier::make();

    $count = 0;
    $N = 4;

    foreach (range(1, $N) as $i) {
        Coroutine::create(function () use ($barrier, &$count) {
//            Coroutine::sleep(0.5);
            $count++;
        });
    }
    Barrier::wait($barrier, 1);
    var_dump($count, $N);
});
```